File: build-web.js
#!/usr/bin/env node
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const root = __dirname;
const flutterDir = path.join(root, '.vercel-flutter');
const flutterBin = process.platform === 'win32'
  ? path.join(flutterDir, 'bin', 'flutter.bat')
  : path.join(flutterDir, 'bin', 'flutter');
const pubspecPath = path.join(root, 'pubspec.yaml');
const webPubspecPath = path.join(root, 'pubspec_web.yaml');
const backupPath = path.join(root, 'pubspec.yaml.bak');

function run(command) {
  execSync(command, { cwd: root, stdio: 'inherit' });
}

function quote(value) {
  return `"${value}"`;
}

function restorePubspec() {
  if (fs.existsSync(backupPath)) {
    fs.copyFileSync(backupPath, pubspecPath);
    fs.unlinkSync(backupPath);
  }
}

function installFlutterIfNeeded() {
  if (fs.existsSync(flutterBin)) {
    return;
  }

  if (fs.existsSync(flutterDir)) {
    throw new Error(
      `${flutterDir} exists but ${flutterBin} is missing. Delete .vercel-flutter and rebuild.`,
    );
  }

  console.log('Installing Flutter stable for this build...');
  run(`git clone https://github.com/flutter/flutter.git -b stable --depth 1 ${quote(flutterDir)}`);
}

function writeBuildStamp() {
  const commit = process.env.VERCEL_GIT_COMMIT_SHA
    || execSync('git rev-parse HEAD', { cwd: root }).toString().trim();
  const branch = process.env.VERCEL_GIT_COMMIT_REF
    || execSync('git branch --show-current', { cwd: root }).toString().trim();
  const stamp = {
    commit,
    branch,
    builtAt: new Date().toISOString(),
  };

  fs.mkdirSync(path.join(root, 'build', 'web'), { recursive: true });
  fs.writeFileSync(
    path.join(root, 'build', 'web', 'build-stamp.json'),
    `${JSON.stringify(stamp, null, 2)}\n`,
  );
  console.log(`Build stamp: ${branch} @ ${commit.slice(0, 7)}`);
}

console.log('Building Safe Mother Malawi web app...');

try {
  if (!fs.existsSync(webPubspecPath)) {
    throw new Error('pubspec_web.yaml is missing.');
  }

  installFlutterIfNeeded();

  run(`${quote(flutterBin)} config --no-analytics`);

  console.log('Using web pubspec...');
  restorePubspec();
  fs.copyFileSync(pubspecPath, backupPath);
  fs.copyFileSync(webPubspecPath, pubspecPath);

  console.log('Installing Flutter dependencies...');
  run(`${quote(flutterBin)} pub get`);

  console.log('Building Flutter web release...');
  run(`${quote(flutterBin)} build web --release --dart-define=FLUTTER_WEB_USE_SKIA=true`);

  restorePubspec();
  writeBuildStamp();

  console.log('Web build completed successfully.');
} catch (error) {
  restorePubspec();
  console.error(`Build failed: ${error.message}`);
  process.exit(1);
}

File: build_web.py
#!/usr/bin/env python3
"""
Build script for web platform
Removes firebase_messaging from pubspec.yaml before building for web
"""

import os
import sys
import subprocess
import shutil
import re

def remove_firebase_messaging(pubspec_path):
    """Remove firebase_messaging from pubspec.yaml"""
    with open(pubspec_path, 'r') as f:
        content = f.read()
    
    # Remove firebase_messaging and its configuration
    # Match: firebase_messaging: ^14.6.0 or firebase_messaging: \n    version: ...
    pattern = r'  firebase_messaging:.*?(?=\n  [a-z_]+:|$)'
    content = re.sub(pattern, '', content, flags=re.DOTALL)
    
    with open(pubspec_path, 'w') as f:
        f.write(content)
    
    print("✓ Removed firebase_messaging from pubspec.yaml")

def restore_pubspec(pubspec_path, backup_path):
    """Restore original pubspec.yaml"""
    shutil.copy(backup_path, pubspec_path)
    print("✓ Restored original pubspec.yaml")

def run_command(cmd, description):
    """Run a shell command"""
    print(f"\n→ {description}...")
    result = subprocess.run(cmd, shell=True)
    if result.returncode != 0:
        print(f"✗ {description} failed!")
        sys.exit(1)
    print(f"✓ {description} completed")

def main():
    print("Building Safe Mother Malawi for Web...\n")
    
    pubspec_path = "pubspec.yaml"
    backup_path = "pubspec.yaml.bak"
    
    try:
        # Install Flutter if not already installed
        if not os.path.exists("flutter/bin/flutter"):
            run_command(
                "git clone https://github.com/flutter/flutter.git -b stable --depth 1",
                "Cloning Flutter"
            )
            run_command(
                "flutter/bin/flutter config --no-analytics",
                "Configuring Flutter"
            )
        
        # Backup original pubspec.yaml
        shutil.copy(pubspec_path, backup_path)
        print("✓ Backed up pubspec.yaml")
        
        # Remove firebase_messaging
        remove_firebase_messaging(pubspec_path)
        
        # Run flutter pub get
        run_command("flutter/bin/flutter pub get", "Flutter pub get")
        
        # Build for web
        run_command(
            "flutter/bin/flutter build web --release --dart-define=FLUTTER_WEB_USE_SKIA=true",
            "Flutter web build"
        )
        
        print("\n✓ Web build completed successfully!")
        
    except Exception as e:
        print(f"\n✗ Error: {e}")
        sys.exit(1)
    finally:
        # Always restore original pubspec.yaml
        if os.path.exists(backup_path):
            restore_pubspec(pubspec_path, backup_path)
            os.remove(backup_path)

if __name__ == "__main__":
    main()
